src/public/components/router.js

```diff
@@ -1,6 +1,83 @@
-import K, * as U from "karet.util"
-import React     from "karet"
+import * as R     from "ramda"
+import * as L     from "partial.lenses"
+import K, * as U  from "karet.util"
+import React      from "karet"
+import toRegex    from 'path-to-regexp'
 
-export default U.withContext(({routes, NotFound}, {path}) => U.fromKefir(K(path, path =>
-  React.createElement(routes[path] || NotFound)
-)))
+const W = a => b => a( b )( b )
+
+// TODO: Make it possible to add routes more than once?
+
+// begin routes initialization
+const partitionDynamicAndStatic =
+  R.partition( R.compose( R.lt( 0 )
+                        , R.length
+                        , L.get( [ 'regex', 'keys' ] )
+                        )
+             )
+
+const addRegexes =
+  R.map( W( R.compose( L.set( 'regex' )
+                     , toRegex
+                     , L.get( 'route' )
+                     )
+          )
+       )
+
+const expandRoutes =
+  R.compose( R.map( R.zipObj( [ 'route', 'component' ] ) )
+           , R.toPairs
+           )
+
+const init =
+  R.compose( R.zipObj( [ 'dynamic', 'static' ] )
+           , partitionDynamicAndStatic
+           , addRegexes
+           , expandRoutes
+           )
+// end routes initialization
+
+//
+
+// begin route resolution
+const resolveStatic = path =>
+  R.compose( R.find( R.propEq( 'route', path ) )
+           , L.get( 'static' )
+           )
+
+const prepParams =
+  R.compose( R.zipObj
+           , R.map( L.get( 'name' ) )
+           )
+
+const resolveDynamic = path =>
+  R.compose( R.reduce( ( acc, { component, route, regex } ) =>
+                       { const match = regex.exec( path )
+                         return match !== null
+                                ? R.reduced( { component, path, route, regex
+                                             , params: prepParams( regex.keys )( R.tail( match ) )
```

Comment:
Here I would suggest avoiding collecting the params into a `params` property and instead passing the parameters as direct, unnested, props to the component.

Note that `path` is available from the context.  I probably wouldn't pass it as a prop to the component at all.

Hmm... Do you have a specific use case in mind for passing the `route` and `regex` props to the component?  I'm not sure that the `route` and `regex` are useful props for the component.  I would probably just drop those.

So, perhaps something like this:

```js
const resolveDynamic = path =>
  R.compose( R.reduce( ( acc, { component, regex } ) =>
                        { const match = regex.exec( path )
                          return match !== null
                                 ? R.reduced( { component
                                              , props: prepParams( regex.keys )( R.tail( match ) } )
// ...
```